pyccel/ast/variable.py

```diff
@@ -357,6 +351,14 @@ def is_ndarray(self):
     def __str__(self):
         return str(self.name)
 
+    def __eq__(self, other):
+        if isinstance(other, Variable):
+            return self._name == other.name
+        return False
+
+    def __hash__(self):
+        return hash(type(self).__name__+ self._name)
```

Comment:
The `__eq__` and `__hash__` don't match. A `ValuedVariable` is a `Variable` but it will not have the same hash as a `Variable` with the same name

Reply:
I totally forgot about valuedVariable

Reply:
`TupleVariable` and `DottedVariable` too. Especially `DottedVariable`, `cls.a`=/=`a`

Reply:
do you think a ```DottedVariable``` should have its own``` __eq__ ```?

Reply:
Actually that's a really good idea. Currently we just test the name. I'm not sure if the name of `a.b` is `'b'` or `'a.b'` if it's the first then we have a problem, if it's the second then it's a strange test for things such as `f().b`